### PR TITLE
Generate: validate `model_kwargs` (and catch typos in generate arguments)

### LINKED ISSUE #18130: model.generate doesn't validate kwargs
### Feature request

I made a mistake in a script:
```
model.generate(**tokens, in_length=num_tokens)
```
missing `m` in `min_length` and I was puzzling over why I was getting unexpected results (as it was using the default value which was quite different from mine)

Would it be possible to have `generate` validate its input and assert on unexpected args?

Thank you!

@patrickvonplaten

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

A common cause for issues in generate is around it not behaving as expected, as arguments can be silently ignored as part of the selected generation submethod (greedy_search, sample, ...). Typos also often fly under the radar, as the method accepts **model_kwargs, which in turn are passed to models that also accept **kwargs.

This PR solves the low-hanging fruit (derived from https://github.com/huggingface/transformers/pull/18218): validates `model_kwargs`, which notifies the users about problems in model arguments AND about typos, as they will fall in `model_kwargs`.

Will open a PR with TF and FLAX equivalent after this one gets merged. A solution for the other generation arguments is also on the way :) 

Fixes https://github.com/huggingface/transformers/issues/18130

___________________

Here is an example of the output for

```python
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("distilgpt2")
model = AutoModelForCausalLM.from_pretrained("distilgpt2")

prompt = tokenizer(["hello world"], return_tensors="pt")
model.generate(**prompt, do_samples=True, foo="bar")
```

image

### CODE REVIEW

File: src/transformers/generation_utils.py [hunk 1]
```diff
@@ -841,6 +841,34 @@ def compute_transition_beam_scores(
...
+    def _validate_model_kwargs(self, model_kwargs: Dict[str, Any]):
+        """Validates model kwargs for generation. Generate argument typos will also be caught here."""
+        # Excludes arguments that are handled before calling any model function
+        if self.config.is_encoder_decoder:
+            for key in ["decoder_input_ids"]:
+                model_kwargs.pop(key, None)
+
+        # Transfo_XL does not use have "attention_mask" as an argument, and it is harmless (it is being passed in the
+        # tests, through, hence this ad hoc exception)
+        if "transfoxl" in str(self).lower():
```
Comment (sgugger) [inline src/transformers/generation_utils.py:853]: Maybe check for `self.config.__class__.__name__` instead here.

File: src/transformers/generation_utils.py [hunk 2]
```diff
@@ -841,6 +841,34 @@ def compute_transition_beam_scores(
...
+            model_kwargs.pop("attention_mask", None)
+
+        unused_model_args = []
+        model_args = set(inspect.signature(self.prepare_inputs_for_generation).parameters)
+        # `kwargs`` if often used to handle optional forward pass inputs like `attention_mask`. If
```
Comment (sgugger) [inline src/transformers/generation_utils.py:858]:
```suggestion
        # `kwargs` if often used to handle optional forward pass inputs like `attention_mask`. If
```

File: tests/generation/test_generation_utils.py [hunk 3]
```diff
@@ -2699,3 +2699,22 @@ def test_constrained_beam_search_mixin_type_checks(self):
...
+
+        encoder_input_str = "Hello world"
+        input_ids = tokenizer(encoder_input_str, return_tensors="pt").input_ids
+
+        # typos are quickly detected (the correct argument is `do_sample`)
+        with self.assertRaises(ValueError):
+            model.generate(
+                input_ids,
+                do_samples=True,
+            )
```
Comment (sgugger) [inline tests/generation/test_generation_utils.py:2715]:
```suggestion
            model.generate(input_ids, do_samples=True)
```

File: tests/generation/test_generation_utils.py [hunk 4]
```diff
@@ -2699,3 +2699,22 @@ def test_constrained_beam_search_mixin_type_checks(self):
...
+
+    def test_validate_generation_inputs(self):
+        tokenizer = AutoTokenizer.from_pretrained("patrickvonplaten/t5-tiny-random")
+        model = AutoModelForSeq2SeqLM.from_pretrained("patrickvonplaten/t5-tiny-random")
+
+        encoder_input_str = "Hello world"
+        input_ids = tokenizer(encoder_input_str, return_tensors="pt").input_ids
+
+        # typos are quickly detected (the correct argument is `do_sample`)
+        with self.assertRaises(ValueError):
```
Comment (sgugger) [inline tests/generation/test_generation_utils.py:2711]: Can we check that the error contains the right text around do_samples here?

File: tests/generation/test_generation_utils.py [hunk 5]
```diff
@@ -2699,3 +2699,22 @@ def test_constrained_beam_search_mixin_type_checks(self):
...
+            model.generate(
+                input_ids,
+                do_samples=True,
+            )
+
+        # arbitrary arguments that will not be used anywhere are also not accepted
+        with self.assertRaises(ValueError):
```
Comment (sgugger) [inline tests/generation/test_generation_utils.py:2718]: Same here, let's check the error contains the right texts.

**Review by sgugger — COMMENTED** [review]
Thanks for working on this! This should also be applied to the TensorFlow and Flax generate functions no? (Can be done in followup PRs of course.)

File: src/transformers/generation_utils.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/generation_utils.py:853]: Think I'd be in favor of displaying the normal "this is not used" error message also for Transfo-XL attention_mask even if we use it in the test -> maybe we should update the tests then instead. If Transfo-XL doesn't have `attention_mask` and the user passes it we should state that it is indeed not used no?

**Review by patrickvonplaten — APPROVED** [review]
Thanks a lot for working on this - this looks very nice to me! Happy to start just with PT and then do TF and Flax in a follow-up PR

File: src/transformers/generation_utils.py (hunk 1 above)
Comment (gante) [reply src/transformers/generation_utils.py]: (removed this exception, will fix tests correspondingly)

**Comment by gante** [comment]
@sgugger ready for a final review :)

**Review by sgugger — APPROVED** [review]
Thanks again for all your work on this!

**Comment by stas00** [comment]
Super-useful, thank you, @gante!

**Pull request merged**